uaclient/jobs/eol_status.py

```diff
@@ -0,0 +1,54 @@
+import os
+from typing import Dict
+
+from uaclient import messages, util
+from uaclient.config import UAConfig
+
+
+def get_os_release_vars() -> Dict:
+    vars = dict()
+    content = util.load_file("/etc/os-release")
+    for line in content.splitlines():
+        k, v = line.rstrip().split("=")
+        vars[k] = v.strip('"')
+    return vars
+
+
+def check_eol_and_update(cfg: UAConfig) -> bool:
+    if cfg.is_attached:
+        False
+    os_vars = get_os_release_vars()
+    version_id = os_vars.get("VERSION_ID", "NO-VERSION_ID")
+    codename = os_vars.get(
+        "UBUNTU_CODENAME", "NO-UBUNTU_CODENAME-" + version_id
+    )
+    to_eol, err = util.subp(
+        ["ubuntu-distro-info", "--series", codename, "-yeol"]
+    )
+    days_to_eol = int(to_eol)
+    if days_to_eol >= 1:
+        return True
+    apt_suite = codename + "-" + "infra"
+    apt_origin = "UbuntuESM"
+    apt_pref_file = "/etc/apt/preferences.d/ubuntu-esm-infra"
+    apt_source_file = "/etc/apt/sources.list.d/ubuntu-esm-infra.list"
+    apt_key = "ubuntu-advantage-esm-infra-trusty.gpg"
+    if not os.path.exists("/etc/apt/trusted.gpg.d/" + apt_key):
+        src = "/usr/share/keyrings/" + apt_key
+        dest = "/etc/apt/trusted.gpg.d/" + apt_key
+        util.subp(["cp", src, dest])
+
+    if not os.path.exists(apt_source_file):
+        util.write_file(
+            apt_source_file,
+            messages.APT_SOURCE_ESM_CONTENT.format(
+                service="infra", apt_suite=apt_suite
+            ),
+        )
+        util.write_file(
+            apt_pref_file,
+            messages.UNATTACHED_APT_ESM_PREF_CONTENT.format(
+                apt_origin=apt_origin, release=codename
+            ),
+        )
```

Comment:
We already do some similar things in `uaclient.apt.add_auth_apt_repo` which includes a function for the gpg key.

repo file paths already have string templates defined on the `RepoEntitlement` class which can be reused here.

I'd guess this implementation will end up looking something like this:
```suggestion
    ESMInfraEntitlement().setup_unauthenticated_repo()
    ESMAppsEntitlement().setup_unauthenticated_repo()
```

Where `setup_unauthenticated_repo` is defined on `ESMBaseEntitlement` and looks something like:
```python
class ESMBaseEntitlement(repo.RepoEntitlement):
    ....existing stuff
    def setup_unauthenticated_repo(self):
        apt.setup_unauthenticated_repo(
            repo_file_name=self.repo_list_file_tmpl.format(name=self.name),
            repo_url="https://esm.ubuntu.com/{service}".format(service=self.apt_repo_name),
            suites=["{series}-updates", "{series}-security"]
            keyfile=self.repo_key_file
        )
```

and apt.setup_unauthenticated_repo is loosely based on the existing code in apt.add_auth_apt_repo